README.md
# Dynamic Illumination
Small Module for Foundry VTT that adds addtional lighting buttons for Dawn/Dusk, and ties Global Illumination to the scene Darkness level.
***
## Features
### Dawn and Dusk
Adds additional buttons to the Lighting toolbar to set the time to Dawn or Dusk
***  
### Time of Day Color and Darkness Level
Allows each time of day (Dawn/Day/Dusk/Night) to have a different scene color and darkness level.  
##### Using the buttons as GM  
<img src="images/dynamic-illumination.gif" height=200>

***  
### Dynamic Global Illumination
Optionally provides a darkness threshold at which Global Illumination will be disabled.  
_Note:_ Scene reload occurs when Global Illumination changes

##### Player view of Time Changes
<img src="images/dynamic-illumination-player.gif" height=200>

***
#### Donations
I have a patreon if you'd like to donate. You shouldn't, but I won't stop you:
https://www.patreon.com/delVhar

CHANGELOG.md
# Dynamic Illumination Changelog
***
## 0.2.4
* Fixed a bug that would leave the darkness color stuck if the module was removed or disabled.

***
## 0.2.2
* Added experimental feature to interpolate scene color instead of setting it. Seems to work ok, but does send a lot of scene updates.
* Fixed bug where color changes didn't occur correctly on button press.

***
## 0.2.0
* Added option to enable darkness level transitions (transition will still be instant if Global Illumination changes)
* Added configurable delay to color change during animated transitions (transition will still be instant if Global Illumination changes)
    * This is a pretty basic implementation using setTimeout, please raise an issue if you see any problems with this.
    * The color change always occurs first if the darkness level is currently set to 0.
* Moved Global Illumination change to occur on the 'updateScene' hook
    * This will mean global illumination change can be triggered by other modules that alter the scene darkness
* General code clean up - moved transitions into a shared function for neatness
* Replaced all cases of 'illumination' with 'dynamic-illumination' for consistency in settings/file names etc.

*Notes*  
I still haven't worked out why the module settings sometimes show under 'General Module Settings'. Looking at the foundry source code, that should only occur if the module title isn't set, which it is...  
The behaviour seems inconsistent, I suspect it's because I changed the title at some point. It might be worth uninstalling and reinstalling the module if you run into the issue.
***
## 0.1.11
* Renamed to Dynamic Illumination
* Updated README

***
## 0.1.10
* Initial version added to FoundryVTT modules list. Versions older than this alternated between working in a way I didn't want, or broken to hell :D

***
#### Donations
I have a patreon if you'd like to donate. You shouldn't, but I won't stop you:
https://www.patreon.com/delVhar


scripts/illumination.js
function toggleGlobalLight(darkness)
{
    //console.log(`Changing darkness to ${darkness}. GI threshold is ${game.settings.get("illumination","darknessThreshold")}. Current GI is: ${canvas.scene.data.globalLight}`);
    if(game.settings.get("illumination","linkGlobalLight"))
    {
        if(darkness >= game.settings.get("illumination","darknessThreshold"))
        {
            if(canvas.scene.data.globalLight)
            {
                return {globalLight: false}
            }
        }
        else if(!canvas.scene.data.globalLight)
        {
            return {globalLight: true}
        }
    }
    return {}
}

Hooks.on('getSceneControlButtons', controls => {
    let control = controls.find(c => c.name === "lighting") || controls[5];

    // Day Button override
    control.tools[1].onClick = () =>
    {
        var level=game.settings.get("illumination","dayLevel");
        canvas.scene.update({darkness: level}, {animateDarkness: true})
        var global = toggleGlobalLight(level);
        console.log(global);
        if(global.hasOwnProperty("globalLight"))
            canvas.scene.update(global)     
        canvas.scene.setFlag("core","darknessColor", game.settings.get("illumination","dayColor"))
    }

    // Night Button Override
    control.tools[2].onClick = () =>
    {
        var level=game.settings.get("illumination","nightLevel");
        canvas.scene.update({darkness: level}, {animateDarkness: true})
        var global = toggleGlobalLight(level);
        console.log(global);
        if(global.hasOwnProperty("globalLight"))
            canvas.scene.update(global)                   
        canvas.scene.setFlag("core","darknessColor", game.settings.get("illumination","nightColor"))
    }

    control.tools.splice(2, 0, {
		name: "dusk",
		title: "Transition to Dusk",
		icon: "far fa-moon",
		visible: game.settings.get("illumination", "showDawnDusk"),
		onClick: () => {
            var level=game.settings.get("illumination","duskLevel");
            canvas.scene.update({darkness: level}, {animateDarkness: true})
            var global = toggleGlobalLight(level);
            console.log(global);
            if(global.hasOwnProperty("globalLight"))
                canvas.scene.update(global)
            canvas.scene.setFlag("core","darknessColor", game.settings.get("illumination","duskColor"))
        }
	});
    control.tools.splice(1, 0, {
		name: "dawn",
		title: "Transition to Dawn",
		icon: "far fa-sun",
		visible: game.settings.get("illumination", "showDawnDusk"),
		onClick: () => {
            var level=game.settings.get("illumination","dawnLevel");
            canvas.scene.update({darkness: level}, {animateDarkness: true})
            var global = toggleGlobalLight(level);
            console.log(global);
            if(global.hasOwnProperty("globalLight"))
                canvas.scene.update(global)
            canvas.scene.setFlag("core","darknessColor", game.settings.get("illumination","dawnColor"));
        }
	});

});

Hooks.once("init", () => {

    game.settings.register("illumination", "linkGlobalLight", {
		name: game.i18n.localize("illumination.linkGlobalLight.name"),
		hint: game.i18n.localize("illumination.linkGlobalLight.hint"),
		scope: "world",
		config: true,
		default: true,
		type: Boolean
    });
    if(game.settings.get("illumination", "linkGlobalLight"))
    {
        game.settings.register("illumination", "darknessThreshold", {
            name: game.i18n.localize("illumination.darknessThreshold.name"),
            hint: game.i18n.localize("illumination.darknessThreshold.hint"),
            scope: "world",
            config: true,
            default: 1.0,
            type: Number,
            range: {min: 0.0, max: 1.0, step: 0.05}
        });
    }
    game.settings.register("illumination", "showDawnDusk", {
		name: game.i18n.localize("illumination.showDawnDusk.name"),
		hint: game.i18n.localize("illumination.showDawnDusk.hint"),
		scope: "world",
		config: true,
		default: true,
		type: Boolean
    });

    game.settings.register("illumination", "dawnColor", {
        name: game.i18n.localize("illumination.dawnColor.name"),
        hint: game.i18n.localize("illumination.dawnColor.hint"),
        scope: "world",
        config: true,
        default: "#db9f6d",
        type: String
    });
    

    game.settings.register("illumination", "dawnLevel", {
        name: game.i18n.localize("illumination.dawnLevel.name"),
        hint: game.i18n.localize("illumination.dawnLevel.hint"),
        scope: "world",
        config: true,
        default: 0.75,
        type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

    game.settings.register("illumination", "dayColor", {
		name: game.i18n.localize("illumination.dayColor.name"),
		hint: game.i18n.localize("illumination.dayColor.hint"),
		scope: "world",
		config: true,
		default: "#FFFEFE",
		type: String
    });
    
    game.settings.register("illumination", "dayLevel", {
		name: game.i18n.localize("illumination.dayLevel.name"),
		hint: game.i18n.localize("illumination.dayLevel.hint"),
		scope: "world",
		config: true,
		default: 0,
		type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });
    
    game.settings.register("illumination", "duskColor", {
        name: game.i18n.localize("illumination.duskColor.name"),
        hint: game.i18n.localize("illumination.duskColor.hint"),
        scope: "world",
        config: true,
        default: "#ae6b6b",
        type: String
    });

    game.settings.register("illumination", "duskLevel", {
        name: game.i18n.localize("illumination.duskLevel.name"),
        hint: game.i18n.localize("illumination.duskLevel.hint"),
        scope: "world",
        config: true,
        default: 0.75,
        type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

    game.settings.register("illumination", "nightColor", {
		name: game.i18n.localize("illumination.nightColor.name"),
		hint: game.i18n.localize("illumination.nightColor.hint"),
		scope: "world",
		config: true,
		default: "#3c3351",
		type: String
    }); 

    game.settings.register("illumination", "nightLevel", {
		name: game.i18n.localize("illumination.nightLevel.name"),
		hint: game.i18n.localize("illumination.nightLevel.hint"),
		scope: "world",
		config: true,
		default: 1,
		type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });
    
});
console.log("Illumination is loaded.")

scripts/dynamic-illumination.js
function toggleGlobalLight(darkness)
{
    //console.log(`Changing darkness to ${darkness}. GI threshold is ${game.settings.get("dynamic-illumination","darknessThreshold")}. Current GI is: ${canvas.scene.data.globalLight}`);
    if(game.settings.get("dynamic-illumination","linkGlobalLight"))
    {
        if(darkness >= game.settings.get("dynamic-illumination","darknessThreshold"))
        {
            if(canvas.scene.data.globalLight)
            {
                return {globalLight: false}
            }
        }
        else if(!canvas.scene.data.globalLight)
        {
            return {globalLight: true}
        }
    }
    return {}
}

function changeLighting(level, color)
{
    if(game.settings.get("dynamic-illumination","animateDarknessChange") && !toggleGlobalLight(level).hasOwnProperty('globalLight'))
    {
        if(game.settings.get("dynamic-illumination","interpolateColor")) // Use the color interpolation
        {
            canvas.scene.update({darkness: level}, {animateDarkness: true}).then(() => {
                interpolateSceneColor(color)
            })
        }
        else if(canvas.scene.data.darkness == 0) // We are at 0, so first change color then get darker.
        {
            canvas.scene.setFlag("dynamic-illumination","darknessColor", color).then(()=> {
                canvas.scene.update({darkness: level}, {animateDarkness: true}).then(() => {
                    CONFIG.Canvas.darknessColor = color;
                    canvas.getLayer("LightingLayer").update();
                });            
            });
        }        
        else // Delay the color change as per setting
        {
            canvas.scene.update({darkness: level}, {animateDarkness: true, diff: false}).then(
                setTimeout(() => {
                    //canvas.scene.setFlag("core","darknessColor", color)
                    canvas.scene.setFlag("dynamic-illumination","darknessColor", color).then(()=> {
                        CONFIG.Canvas.darknessColor = color;
                        canvas.getLayer("LightingLayer").update();
                    });
                },game.settings.get("dynamic-illumination","animationColorChangeDelay") * 1000)        
            );
        }
    }
    else
    {
        canvas.scene.setFlag("dynamic-illumination","darknessColor", color).then(()=> {
            canvas.scene.update({darkness: level}, {animateDarkness: false}).then(() => {
                CONFIG.Canvas.darknessColor = color;
                canvas.getLayer("LightingLayer").update();
            });            
        });
    }
}

async function interpolateSceneColor(target="#FFFEFF")
{    
    const interpolationData = [{
        parent: {interpolationSteps: 0},
        attribute: "interpolationSteps",
        to: 20
    }];
    return CanvasAnimation.animateLinear(interpolationData, {
        name: "lighting.darknessColor",
        duration: game.settings.get("dynamic-illumination","animationColorChangeDelay") * 1000,
        ontick: (dt, attributes) => {
            color = interpolateColor(canvas.scene.getFlag("dynamic-illumination","darknessColor"), target, attributes[0].parent.interpolationSteps/attributes[0].to)
            // Only update if we actually changed color
            if(color.toLowerCase() != canvas.scene.getFlag("dynamic-illumination","darknessColor").toLowerCase())
            {                
                canvas.scene.setFlag("dynamic-illumination","darknessColor", color).then(()=> {
                    CONFIG.Canvas.darknessColor = color;
                    canvas.getLayer("LightingLayer").update();
                });
            } 
        }
    }).then(() => {
         //Set it to the target at the end in case it wasn't there for some reason
         canvas.scene.setFlag("dynamic-illumination","darknessColor", color).then(()=> {
            CONFIG.Canvas.darknessColor = color;
            canvas.getLayer("LightingLayer").update();
        });
    }); }

function interpolateColor(color1, color2, factor) {
    if (arguments.length < 3) { 
        factor = 0.5; 
    }
    var c1 = convertHexRGB(color1);
    var c2 = convertHexRGB(color2);
    result=c1;
    for (var i = 0; i < 3; i++) {
        result[i] = Math.round(result[i] + factor * (c2[i] - c1[i]));
    }
    return convertRGBHex(result[0], result[1], result[2]);
};

 // Converts hex string to RGB array 
const convertHexRGB = hex => hex.replace(/^#?([a-f\d])([a-f\d])([a-f\d])$/i,(m, r, g, b) => '#' + r + r + g + g + b + b)
    .substring(1).match(/.{2}/g)
    .map(x => parseInt(x, 16));
// Converts RGB array to hex string  
const convertRGBHex = (r, g, b) => '#' + [r, g, b]
    .map(x => x.toString(16).padStart(2, '0')).join('');

Hooks.on('updateScene', (scene, change, diff, token) => {
    if(change.hasOwnProperty('darkness'))
    {
        var global = toggleGlobalLight(change.darkness);
        if(global.hasOwnProperty("globalLight"))
            scene.update(global)
    }
})

Hooks.on('getSceneControlButtons', controls => {
    let control = controls.find(c => c.name === "lighting");
    if(control == undefined)
        return;

    // Day Button override
    let dayButton = control.tools.findIndex(t => t.name === "day")
    if(dayButton)
    {
        control.tools[dayButton].onClick = () => {changeLighting(game.settings.get("dynamic-illumination","dayLevel"), game.settings.get("dynamic-illumination","dayColor"))};
        control.tools.splice(dayButton, 0, {
            name: "dawn",
            title: "Transition to Dawn",
            icon: "far fa-sun",
            visible: game.settings.get("dynamic-illumination", "showDawnDusk"),
            onClick: () => {changeLighting(game.settings.get("dynamic-illumination","dawnLevel"), game.settings.get("dynamic-illumination","dawnColor"))}
        });
    }
    else
    {
        console.error("Unable to locate button for 'Day' in lighting toolbar.");
    }

    // Night Button Override
    let nightButton = control.tools.findIndex(t => t.name === "night")
    if(nightButton)
    {
        control.tools[nightButton].onClick = () => {changeLighting(game.settings.get("dynamic-illumination","nightLevel"), game.settings.get("dynamic-illumination","nightColor"))};
        control.tools.splice(nightButton, 0, {
            name: "dusk",
            title: "Transition to Dusk",
            icon: "far fa-moon",
            visible: game.settings.get("dynamic-illumination", "showDawnDusk"),
            onClick: () => {changeLighting(game.settings.get("dynamic-illumination","duskLevel"), game.settings.get("dynamic-illumination","duskColor"))}
        });
    }
    else
    {
        console.error("Unable to locate button for 'Night' in lighting toolbar.");
    }

});

Hooks.once("init", () => 
{
    game.settings.register("dynamic-illumination", "linkGlobalLight", {
		name: game.i18n.localize("dynamic-illumination.linkGlobalLight.name"),
		hint: game.i18n.localize("dynamic-illumination.linkGlobalLight.hint"),
		scope: "world",
		config: true,
		default: true,
		type: Boolean
    });

    game.settings.register("dynamic-illumination", "darknessThreshold", {
        name: game.i18n.localize("dynamic-illumination.darknessThreshold.name"),
        hint: game.i18n.localize("dynamic-illumination.darknessThreshold.hint"),
        scope: "world",
        config: true,
        default: 1.0,
        type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

    game.settings.register("dynamic-illumination", "animateDarknessChange", {
		name: game.i18n.localize("dynamic-illumination.animateDarknessChange.name"),
		hint: game.i18n.localize("dynamic-illumination.animateDarknessChange.hint"),
		scope: "world",
		config: true,
		default: true,
		type: Boolean
    });

    game.settings.register("dynamic-illumination", "interpolateColor", {
		name: game.i18n.localize("dynamic-illumination.interpolateColor.name"),
		hint: game.i18n.localize("dynamic-illumination.interpolateColor.hint"),
		scope: "world",
		config: true,
		default: false,
		type: Boolean
    });

    game.settings.register("dynamic-illumination", "animationColorChangeDelay", {
		name: game.i18n.localize("dynamic-illumination.animationColorChangeDelay.name"),
		hint: game.i18n.localize("dynamic-illumination.animationColorChangeDelay.hint"),
		scope: "world",
		config: true,
		default: 7.5,
        type: Number,
        range: {min: 0.0, max: 10.0, step: 0.5}
    });

    game.settings.register("dynamic-illumination", "showDawnDusk", {
		name: game.i18n.localize("dynamic-illumination.showDawnDusk.name"),
		hint: game.i18n.localize("dynamic-illumination.showDawnDusk.hint"),
		scope: "world",
		config: true,
		default: true,
		type: Boolean
    });

    game.settings.register("dynamic-illumination", "dawnColor", {
        name: game.i18n.localize("dynamic-illumination.dawnColor.name"),
        hint: game.i18n.localize("dynamic-illumination.dawnColor.hint"),
        scope: "world",
        config: true,
        default: "#db9f6d",
        type: String
    });
    

    game.settings.register("dynamic-illumination", "dawnLevel", {
        name: game.i18n.localize("dynamic-illumination.dawnLevel.name"),
        hint: game.i18n.localize("dynamic-illumination.dawnLevel.hint"),
        scope: "world",
        config: true,
        default: 0.75,
        type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

    game.settings.register("dynamic-illumination", "dayColor", {
		name: game.i18n.localize("dynamic-illumination.dayColor.name"),
		hint: game.i18n.localize("dynamic-illumination.dayColor.hint"),
		scope: "world",
		config: true,
		default: "#FFFEFE",
		type: String
    });
    
    game.settings.register("dynamic-illumination", "dayLevel", {
		name: game.i18n.localize("dynamic-illumination.dayLevel.name"),
		hint: game.i18n.localize("dynamic-illumination.dayLevel.hint"),
		scope: "world",
		config: true,
		default: 0,
		type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });
    
    game.settings.register("dynamic-illumination", "duskColor", {
        name: game.i18n.localize("dynamic-illumination.duskColor.name"),
        hint: game.i18n.localize("dynamic-illumination.duskColor.hint"),
        scope: "world",
        config: true,
        default: "#ae6b6b",
        type: String
    });

    game.settings.register("dynamic-illumination", "duskLevel", {
        name: game.i18n.localize("dynamic-illumination.duskLevel.name"),
        hint: game.i18n.localize("dynamic-illumination.duskLevel.hint"),
        scope: "world",
        config: true,
        default: 0.75,
        type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

    game.settings.register("dynamic-illumination", "nightColor", {
		name: game.i18n.localize("dynamic-illumination.nightColor.name"),
		hint: game.i18n.localize("dynamic-illumination.nightColor.hint"),
		scope: "world",
		config: true,
		default: "#3c3351",
		type: String
    }); 

    game.settings.register("dynamic-illumination", "nightLevel", {
		name: game.i18n.localize("dynamic-illumination.nightLevel.name"),
		hint: game.i18n.localize("dynamic-illumination.nightLevel.hint"),
		scope: "world",
		config: true,
		default: 1,
		type: Number,
        range: {min: 0.0, max: 1.0, step: 0.05}
    });

})

Hooks.once("canvasInit", () => {
    canvas.scene.unsetFlag("core","darknessColor");  // Delete darknessColor Flag to clean up old usage.
    // Set Canvas Darkness color to match flag
    var color = canvas.scene.getFlag("dynamic-illumination","darknessColor");
    if(color == undefined)
    {
        canvas.scene.setFlag("dynamic-illumination","darknessColor", "#110033");
        color = "#110033";
    }
    CONFIG.Canvas.darknessColor = color;    
});
